### DISCUSSION TITLE
Implementation of Corda Branch

### DISCUSSION THREAD

**Discussion by ami000** [TradeTrust/tradetrust-website#324] [General]
Hi Team, 

Creating a discussion here as a precursor to our PR. (reference from Shujing LIEW)
cc: @alokrajiv

We at Trames have implemented a Corda Implementation of TradeTrust documents and this PR is regarding the validator functionality for Corda TT documents going in as a part of the main tradetrust website.
We have mirrored the existing repo styles which also pass your lint tests. But we have some questions pertinent to the PR itself which are as follows:
- Are there any specific contribution guidelines in a `CONTRIBUTING.md` or general guidelines in the Netlify/Readme.md, because we couldn't find it on master or on wiki.
- The current `master` branch does not match the `corda/dev` branch. Would you like us to merge?

Once we get clarity on these, we will open the PR to `corda/dev` branch from our repo soon.
The sections below outline some initial discussions.

#### Motivation

While the exchange of electronic documents can be achieved in a secure and tamper-proof way using public blockchains, public visibility and scalability are some of the issues faced by most governments and enterprises. We at Trames using a P2P distributed ledger model came up with a TradeTrust implementation where the electronic documents are shared with peers on a "need to know" basis thus preserving privacy and also achieve scalability by nodes having to retain data of only the transactions that they are involved in. Corda is used as the underlying blockchain to achieve this.

#### Key Considerations
- All crucial document data such as created date, a hash of contents, the issuer, the owner are part of the ledger state and the actual content of the document as key-value pairs/attachment to the transaction.
- The document can be subjected to actions like `update`, `transfer ownership` etc. to participants across the network while having historic data of all the actions stored in the ledger.
-  Ledger level rules on how the document state can evolve will be enforced before each transaction, thus preserving the ownership, update, transfer rights.
-  For public validation of a document's integrity, a separate validator node holds all the masked documents. The masked data has only crucial document details enough for validation, thus preventing critical document data from being exposed to the public-facing validator node.
-  Verification of the document can be done by comparing the document data integrity with the one on the ledger besides the DNS check and issuer validity.
- Each government/business can host their own node and be part of the network thus giving them true control over their data and giving complete privacy control to the document owner.
- Multiple individual units of the same organisation / smaller organisations can be hosted as accounts in a single node.

#### Proposed - Trames’ Corda Implementation of TradeTrust

This Corda Network is run in the Trames Backend. Setting `NET="Corda Enterprise"` can be used to access it.
We have implemented an open verifier API that calls the Corda flows internally for document verification. The response includes some verification fragments like DNS-verification(dnsVerified: Boolean), issuer identity validation(issuerValid: Boolean), document hash validation(isDocValid: Boolean) and document status(status: Issued/Consumed/Revoked).

The response from the open verifier API will be as shown below: 
        {
            "issuerValid": true,
            "isDocValid": false,
            "status": "Issued",
            "dltID": "601bba52215XXXXXXXXXXXXXX",
            "dnsVerified": true,
            "message": "Document verified successfully."
        }

The TradeTrust website validator page will then display the document according to the response it gets from open verifier API:

| Open Verifier API Response        	| Display Message on TradeTrust Verify Website                           	| Blockchain State                                                                                                                                                                               	|
|-----------------------------------	|------------------------------------------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| isDocValid: false                 	| Document is tampered with                                              	| Document exists in blockchain but calculated current document<br>hash is different from original document hash.                                                                                	|
| issuerValid: false                	| Document issuer is invalid                                             	| Current document issuer is invalid.                                                                                                                                                            	|
| dnsVerified: false                	| DNS verification failed                                                	| DNS-TXT Record Verification Failed                                                                                                                                                             	|
| status: "Issued/Consumed/Revoked" 	| Green/Orange/Red colour on UI to indicate respective document state 	| Issued - Document issuer and DNS are valid.<br>Consumed - Document was issued but is now consumed/surrendered.<br>Revoked - Document was issued but was later revoked by owner. 	|

**Comment by alokrajiv** [comment]
@Nebulis Hi, continuing on https://github.com/TradeTrust/tradetrust-website/issues/322 here, as per your recommendation. 

**Do let us know if you or some of your fellow maintainers can help us answer the 2 questions.** 

We added more information into our implementation as well, so do advise if you need us directly into a .md inside the repo.

**Reply by rjchow** [reply]
Hi, there isn’t a contributors.md right now, just ensure

1. lint passes
2. tests pass
3. new code is tested

we usually use rebase to keep up to date, feel free to use whichever method you’re more comfortable with

**Reply by Nebulis** [reply]
Makes me think that sometimes ci does not trigger on forks.
